python/lsst/daf/butler/core/dimensions/_records.py

```diff
@@ -170,23 +171,22 @@ def direct(
 
         This method should only be called when the inputs are trusted.
         """
-        _recItems = record.items()
+        # This method requires tuples as values of the mapping, but JSON
+        # readers will read things in as lists. Be kind and transparently
+        # transform to tuples
+        _recItems = {k: v if type(v) != list else tuple(v) for k, v in record.items()}  # type: ignore
+
         # Type ignore because the ternary statement seems to confuse mypy
         # based on conflicting inferred types of v.
         key = (
             definition,
-            frozenset((k, v if not isinstance(v, list) else tuple(v)) for k, v in _recItems),  # type: ignore
+            frozenset(_recItems.items()),
```

Comment:
This is could be a pretty inefficient key for spatial dimension records in particular, in which there's a spatial region involved in any comparison.

You really just need to compare the required dimension values, not any of the metadata fields, and since those dimensions are fully specified `definition`,
```
key = (
    definition.name,
    tuple(_recItems[k] for k in definition.required.names),
)
```
is probably pretty close to the best key you can get for dimension records, _if_ the extra effort to extract it doesn't outweigh the advantage of faster comparisons.

Reply:
Same thing here, dimension is not yet anything, it is still an attribute on the serialized record. I could pull de-serializing the definition to before making the key, but it would mean deserializing every definition to create a key to see if there was already a cached record. Couple that with the iteration and I am not sure we would get a speed up, but I can test.

Reply:
oh no I though this is still in the serialized object, I would prefer to not de-serialize anything if it could be avoided. If you thought it safe, I could always drop region

Reply:
FWIW, "deserializing the definition" should just be `univese[dimension]`.  But I agree it's not clear whether it'd be an overall win.

Reply:
but in this case it is in creating the serialized object itself. We are not sure yet what universe to use. There is a universe object that is passed to the `from_simple` method, but that comes later. I could see if this optimization helps there

Reply:
Fair enough.  Sounds like it'd require a fair bit of restructuring if the universe isn't available here.